libs/cardano-ledger-core/src/Cardano/Ledger/PParams.hs

```diff
@@ -0,0 +1,376 @@
+  (Typeable era, ToCBOR (PParamsHKD StrictMaybe era)) => ToCBOR (PParamsUpdate era)
+
+deriving newtype instance
+  (Typeable era, FromCBOR (PParamsHKD StrictMaybe era)) => FromCBOR (PParamsUpdate era)
+
+deriving newtype instance
+  ToJSON (PParamsHKD StrictMaybe era) => ToJSON (PParamsUpdate era)
+
+deriving newtype instance
+  FromJSON (PParamsHKD StrictMaybe era) => FromJSON (PParamsUpdate era)
+
+class
+  ( Era era,
+    Eq (PParamsHKD Identity era),
+    Show (PParamsHKD Identity era),
+    NFData (PParamsHKD Identity era),
+    ToCBOR (PParamsHKD Identity era),
+    FromCBOR (PParamsHKD Identity era),
+    ToJSON (PParamsHKD Identity era),
+    FromJSON (PParamsHKD Identity era),
+    NoThunks (PParamsHKD Identity era),
+    Eq (PParamsHKD StrictMaybe era),
+    Ord (PParamsHKD StrictMaybe era),
+    Show (PParamsHKD StrictMaybe era),
+    NFData (PParamsHKD StrictMaybe era),
+    ToCBOR (PParamsHKD StrictMaybe era),
+    FromCBOR (PParamsHKD StrictMaybe era),
+    NoThunks (PParamsHKD StrictMaybe era)
+  ) =>
+  EraPParams era
+  where
+  -- | Protocol parameters where the fields are represented with a HKD
+  type PParamsHKD (f :: Type -> Type) era = (r :: Type) | r -> era
+
+  -- | Applies a protocol parameters update
+  applyPPUpdates :: PParams era -> PParamsUpdate era -> PParams era
+
+  emptyPParams :: PParams era
+  emptyPParamsUpdate :: PParamsUpdate era
+
+  -- HKD Versions of lenses
+
+  -- | The linear factor for the minimum fee calculation
+  hkdMinFeeAL :: Lens' (PParamsHKD f era) (HKD f Natural)
+
+  -- | The constant factor for the minimum fee calculation
+  hkdMinFeeBL :: Lens' (PParamsHKD f era) (HKD f Natural)
+
+  -- | Maximal block body size
+  hkdMaxBBSizeL :: Lens' (PParamsHKD f era) (HKD f Natural)
+
+  -- | Maximal transaction size
+  hkdMaxTxSizeL :: Lens' (PParamsHKD f era) (HKD f Natural)
+
+  -- | Maximal block header size
+  hkdMaxBHSizeL :: Lens' (PParamsHKD f era) (HKD f Natural)
+
+  -- | The amount of a key registration deposit
+  hkdKeyDepositL :: Lens' (PParamsHKD f era) (HKD f Coin)
+
+  -- | The amount of a pool registration deposit
+  hkdPoolDepositL :: Lens' (PParamsHKD f era) (HKD f Coin)
+
+  -- | epoch bound on pool retirement
+  hkdEMaxL :: Lens' (PParamsHKD f era) (HKD f EpochNo)
+
+  -- | Desired number of pools
+  hkdNOptL :: Lens' (PParamsHKD f era) (HKD f Natural)
+
+  -- | Pool influence
+  hkdA0L :: Lens' (PParamsHKD f era) (HKD f NonNegativeInterval)
+
+  -- | Monetary expansion
+  hkdRhoL :: Lens' (PParamsHKD f era) (HKD f UnitInterval)
+
+  -- | Treasury expansion
+  hkdTauL :: Lens' (PParamsHKD f era) (HKD f UnitInterval)
+
+  -- | Decentralization parameter
+  hkdDL :: Lens' (PParamsHKD f era) (HKD f UnitInterval)
+
+  -- | Extra entropy
+  hkdExtraEntropyL :: Lens' (PParamsHKD f era) (HKD f Nonce)
+
+  -- | Protocol version
+  hkdProtocolVersionL :: Lens' (PParamsHKD f era) (HKD f BT.ProtVer)
+
+  -- | Minimum UTxO value
+  hkdMinUTxOValueL :: Lens' (PParamsHKD f era) (HKD f Coin)
```

Comment:
In order to prevent usage of this lens for eras post Mary era this is what we can do:
```suggestion
  hkdMinUTxOValueL :: AtMostEra MaryEra era => Lens' (PParamsHKD f era) (HKD f Coin)
```

This will cause a compilation error when used in Alonzo onwards.

Actual implementation for the function is still necessary, however we can't just use `error`, because ghc will wrongfully  produce a warning. Therefore we need to use a specially crafted function: `Cardano.Ledger.Core.notSupportedInThisEraL`

If `MaryEra` won't work because it is a type synonym we can  use the actual type: `AtMostEra (ShelleyMAEra 'Mary) era`